Sources/BSON/BSONDecimal128.swift

```diff
@@ -0,0 +1,526 @@
+import Foundation
+import NIO
+
+private extension UInt64 {
+    var upper32bits: Self { self.getBits(0...31) }
+    var lower32bits: Self { self.getBits(32...63) }
+
+    /// Gets this number's bits without shifting the value down to the LSB
+    /// the bits are indexed from MSB at 0 to LSB at 63
+    /// example: (0b11010).getBitsUnshifted(0..<4) == 0b110_00
+    /// (note: the bits before the _ are the one's gotten)
+    func getBitsUnshifted(_ range: ClosedRange<Int>) -> Self {
+        guard range.lowerBound >= 0 else {
+            fatalError("BSONDecimal128: Your range should be bound between [0, 63] was \(range)")
+        }
+        guard range.upperBound <= 63 else {
+            fatalError("BSONDecimal128: Your range should be bound between [0, 63] was \(range)")
+        }
+        var value = UInt64()
+        for i in range {
+            value |= (self & (0b1 << (63 - i)))
+        }
+        return value
+    }
+
+    func getLeastSignificantBits<T: FixedWidthInteger>(_ length: T) -> Self {
+        self.getBits((63 - Int(length))...63)
+    }
+
+    /// Gets this number's bits shifting the value down to the LSB
+    /// the bits are indexed from MSB at 0 to LSB at 63
+    /// example: (0b11010).getBitsUnshifted(0..<4) == 0b110
+    func getBits(_ range: ClosedRange<Int>) -> Self {
+        var value = self.getBitsUnshifted(range)
+        value >>= (63 - range.upperBound)
+        return value
+    }
+
+    func getBit(_ index: Int) -> Self {
+        let shiftAmount = 63 - index
+        let value = (self >> shiftAmount) & 0b1
+        return value
+    }
+
+    mutating func setBit(_ index: Int) {
+        let shiftAmount = 63 - index
+        self |= 1 << shiftAmount
+    }
+}
+
+private extension Array where Element == UInt8 {
+    func decimalDigitsToUInt64() -> UInt64 {
+        var value = UInt64()
+        guard !self.isEmpty else {
+            return value
+        }
+        value = UInt64(self[0])
+        for digit in self[1...] {
+            value *= 10
+            value += UInt64(digit)
+        }
+        return value
+    }
+}
+
+internal struct UInt128: Equatable, Hashable {
+    /// The high order 64 bits
+    internal var hi: UInt64
+    /// The low order 64 bits
+    internal var lo: UInt64
+
+    internal init(hi: UInt64, lo: UInt64) {
+        self.hi = hi
+        self.lo = lo
+    }
+
+    internal init() {
+        self.hi = 0
+        self.lo = 0
+    }
+
+    internal static func multiply(_ left: UInt64, by right: UInt64) -> UInt128 {
```

Comment:
was just reading the UInt64 docs. could we just use this method instead to get the hi and lo values for the UInt128? https://developer.apple.com/documentation/swift/uint64/2885565-multipliedfullwidth

there are also some division related helpers you might be able to use to simplify the division logic

Reply:
I incorporated the multiply helper, the division one proves a bit more tricky with how the logic works around using the hi and lo portions of the number, as well as overwriting the significand with the quotient, would you be okay with keeping our existing logic?

Reply:
yup, that's fine with me! I didn't look at those ones too closely so I wasn't sure if they would help, just thought it might be worth looking into.